source/library/Monadoc/Action/HackageIndex/Process.hs

```diff
@@ -207,36 +207,39 @@ upsertPreference packageName constraint = do
   Monad.void $
     Preference.Upsert.run
       Preference.Preference
-        { Preference.package = Model.key package,
-          Preference.range = Model.key range
+        { Preference.package = package.key,
+          Preference.range = range.key
         }
-  rows <- App.Sql.query "select * from upload inner join version on version.key = upload.version where upload.package = ?" [Model.key package]
+  rows <- App.Sql.query @(Upload.Model Sql.:. Version.Model) "select * from upload inner join version on version.key = upload.version where upload.package = ?" [package.key]
   Monad.forM_ rows $ \(upload Sql.:. version) -> do
-    let isPreferred = Constraint.includes (Version.number $ Model.value version) constraint
-    Monad.when (Upload.isPreferred (Model.value upload) /= isPreferred) $
-      App.Sql.execute "update upload set isPreferred = ? where key = ?" (isPreferred, Model.key upload)
+    let isPreferred = Constraint.includes version.value.number constraint
+    Monad.when (upload.value.isPreferred /= isPreferred) $
+      App.Sql.execute "update upload set isPreferred = ? where key = ?" (isPreferred, upload.key)
 
 updateLatest :: App.App ()
 updateLatest = do
   keys <- Package.Query.getKeys
   Monad.forM_ keys $ \key -> do
     rows <-
-      App.Sql.query
+      App.Sql.query @(Upload.Model Sql.:. Version.Model)
         "select * \
         \ from upload \
         \ inner join version \
         \ on version.key = upload.version \
         \ where upload.package = ?"
         [key :: Package.Key]
-    let f (upload Sql.:. version) =
+    let f ::
+          Upload.Model Sql.:. Version.Model ->
+          Ord.Down (Bool, VersionNumber.VersionNumber, Revision.Revision)
+        f (upload Sql.:. version) =
```

Comment:
Some helper functions are also polymorphic now and require a type signature to pin things down.